pkg/cache/tas_flavor_snapshot.go

```diff
@@ -647,20 +663,37 @@ func (s *TASFlavorSnapshot) sortedDomains(domains []*domain, unconstrained bool)
 func (s *TASFlavorSnapshot) fillInCounts(requests resources.Requests,
 	assumedUsage map[utiltas.TopologyDomainID]resources.Requests,
 	simulateEmpty bool,
-	tolerations []corev1.Toleration) {
+	tolerations []corev1.Toleration,
+	nodeSelectors map[string]string) error {
 	for _, domain := range s.domains {
 		// cleanup the state in case some remaining values are present from computing
 		// assignments for previous PodSets.
 		domain.state = 0
 	}
+	selector, err := labels.ValidatedSelectorFromSet(nodeSelectors)
+	if err != nil {
+		return err
+	}
 	for _, leaf := range s.leaves {
+		// 1. Check Tolerations against Node Taints
 		taint, untolerated := corev1helpers.FindMatchingUntoleratedTaint(leaf.nodeTaints, tolerations, func(t *corev1.Taint) bool {
 			return t.Effect == corev1.TaintEffectNoSchedule || t.Effect == corev1.TaintEffectNoExecute
 		})
 		if untolerated {
 			s.log.V(2).Info("excluding node with untolerated taint", "domainID", leaf.id, "taint", taint)
 			continue
 		}
+		// 2. Check Node Labels against Compiled Selector
+		var nodeLabelSet labels.Set
+		if leaf.nodeLabels != nil {
+			nodeLabelSet = leaf.nodeLabels
+		}
+		// isLowestLevelNode() is necessary because we gather node level information only when
+		// node is the lowest level of the topology
+		if s.isLowestLevelNode() && !selector.Matches(nodeLabelSet) {
```

Comment:
@mimowo @PBundyra FYI I had to add the isLowestLevelNode() condition because we only gather node labels when the lowest level is node. Some tests actually started to fail (`should admit workload which fits in a required topology domain` from `tas_test.go` when this was not the case.

I'm not 100% sure whether it's an indication of some other bug or inconsistency or whether it's expected. PTAL and let me know whether you'd expect this condition here.

Reply:
This is actually a great catch by the tests, uff :). Indeed, we only have the node taints and node labels, if Node is the lowest domain. However, the analogous check was not needed explicitly because for node taints, empty node taints match any Pod. In case of empty node labels, they may not match required Pod selectors.

Reply:
WDYT about ifing the whole part with nodeTaints and nodeSelectors so we only execute this logic when node is the lowest level?

Reply:
Sounds reasonable, but I wouldn't require it if this complicates the PR.

Reply:
We discussed it offline that I would extract the selector compilation out of the `fillInCounts` because it shouldn't be the concern of that function and it also remove the need of error propagation. We may revisit this area of the code later because it looks like it would use some refactoring at some point with a lot of code branching.